crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1340,21 +1365,77 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the length value of a type.
+    ///
+    /// This is used to determine the value that would be returned
+    /// when `len(x)` is called on an object `x`.
+    fn len(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            Type::BytesLiteral(bytes) => return bytes.len(db).try_into().ok(),
+            Type::StringLiteral(string) => {
+                return string.as_str(db).chars().count().try_into().ok()
+            }
+
+            Type::Tuple(tuple) => return tuple.elements(db).len().try_into().ok(),
+
+            ty if !matches!(ty, Type::Instance(..)) => return None,
+
+            _ => {}
```

Comment:
this is incorrect -- we want to treat classes the same as instances here (because a class is an instance of its metaclass). I think you want something like

```rs
match self {
    Type::BytesLiteral(bytes) => return bytes.len(db).try_into().ok(),
    Type::StringLiteral(string) => {
        return string.as_str(db).chars().count().try_into().ok()
    }
    Type::Tuple(tuple) => tuple.elements(db).len().try_into().ok(),
    _ => {}
}

let meta_type = self.to_meta_type();

let Symbol::Type(dunder_len, _) = meta_type.member(db, "__len__") else {
    return None;
};
```

We do something similar here for our `Type::iter()` method; you should be able to look at that for inspiration